packages/kbn-config-schema/src/types/lazy.test.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { Type, schema, lazy } from '../..';
+
+interface RecursiveType {
+  name: string;
+  self: undefined | RecursiveType;
+}
+
+// Test our recursive type inference
+{
+  const id = 'recursive';
+  // @ts-expect-error
+  const testObject: Type<RecursiveType> = schema.object(
+    {
+      name: schema.string(),
+      notSelf: lazy<RecursiveType>(id), // this declaration should fail
+    },
+    { meta: { id } }
+  );
+}
+
+describe('lazy', () => {
+  const id = 'recursive';
+  const object: Type<RecursiveType> = schema.object(
+    {
+      name: schema.string(),
+      self: lazy<RecursiveType>(id),
+    },
+    { meta: { id } }
+  );
```

Comment:
It's a bit verbose, I would have liked to do something [like Zod](https://zod.dev/?id=recursive-types) but with the way `Joi.link` works this just seemed simplest.